File: tests/python/dynamic_programming/maximum_sum_submatrix/test_maximum_sum_submatrix.py
from problems.dynamic_programming.maximum_sum_submatrix.python.maximum_sum_submatrix import \
    maximumSumSubmatrix
from tests.python.utility import load_test_cases


def test_maximum_sum_submatrix() -> None:
    test_cases = load_test_cases("dynamic_programming/maximum_sum_submatrix.json")

    for idx, test in enumerate(test_cases):
        matrix = test["matrix"]
        size = test["size"]
        expected = test["expected"]

        result = maximumSumSubmatrix(matrix, size)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/dynamic_programming/test_maximum_sum_submatrix.cpp
#include "../../../problems/dynamic_programming/maximum_sum_submatrix/cpp/maximum_sum_submatrix.h"
#include "../jsontestbase.h"

#include <vector>

class MaximumSumSubmatrixTest : public JsonTestBase {
public:
  MaximumSumSubmatrixTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/dynamic_programming/maximum_sum_submatrix.json";
  }
};

TEST_F(MaximumSumSubmatrixTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    auto matrix = test["matrix"].get<std::vector<std::vector<int>>>();
    int size = test["size"].get<int>();
    int expected = test["expected"].get<int>();

    int result = maximumSumSubmatrix(matrix, size);

    EXPECT_EQ(result, expected) << "Failed test case: " << test.dump(2);
  }
}
